server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesUpdateProcessor.java

```diff
@@ -156,9 +165,27 @@ private void handleUpdate(MailboxId mailboxId, MailboxUpdateRequest updateReques
         }
    }
 
-    private void checkRole(Optional<Role> role) throws SystemMailboxNotUpdatableException {
-        if (role.map(Role::isSystemRole).orElse(false)) {
-            throw new SystemMailboxNotUpdatableException();
+    private void assertNotSharedOutboxOrDraftMailbox(Mailbox mailbox, MailboxUpdateRequest updateRequest) {
+        mailbox.getRole().ifPresent(role ->
+            {
+                Preconditions.checkArgument(!isSetSharedWithOn(updateRequest.getSharedWith(), mailbox, Role.OUTBOX), "Cannot share 'Outbox'");
+                Preconditions.checkArgument(!isSetSharedWithOn(updateRequest.getSharedWith(), mailbox, Role.DRAFTS), "Cannot share 'Draft'");
+            }
+        );
+    }
+
+    private boolean isSetSharedWithOn(Optional<Rights> rights, Mailbox mailbox, Role role) {
+        return rights.isPresent() && mailbox.hasRole(role);
+    }
+
+    private void checkRoleAndOtherUpdateExceptSharedWith(Mailbox mailbox, MailboxUpdateRequest updateRequest) throws SystemMailboxNotUpdatableException {
```

Comment:
We should use the assert wording.

Moreover, this naming here is clumsy...

I would prefer **assertSystemMailboxesAreNotUpdated**

Reply:
:+1: